massa-execution-worker/src/execution.rs

```diff
@@ -1054,19 +1056,23 @@ impl ExecutionState {
                 *context_guard!(self) = execution_context;
 
                 // run the target function in the bytecode
-                massa_sc_runtime::run_function(
+                let remaining_gas = massa_sc_runtime::run_function(
                     &bytecode,
                     req.max_gas,
                     &target_func,
                     &parameter,
                     &*self.execution_interface,
                 )
                 .map_err(|err| ExecutionError::RuntimeError(err.to_string()))?;
+
+                remaining_gas
             }
-        }
+        };
 
         // return the execution output
-        Ok(context_guard!(self).settle_slot())
+        let mut execution_output = context_guard!(self).settle_slot();
+        execution_output.gas_cost = Some(req.max_gas - remaining_gas);
+        Ok(execution_output)
```

Comment:
Here is an improvement:

```
{
        let mut context = context_guard!(self);
        context.register_used_gas(req.max_gas.saturating_sub(remaining_gas));
        Ok(context.settle_slot())
}
```

Where `register_used_gas` increments the context's `slot_used_gas` with the provided amount of gas. Then `settle_slot` sets the `execution_output.used_gas`.

This should also be used for the normal, non-readonly execute_slot.

Why ? because:
* it avoids carrying around a sense-less used_gas property in the execution outputs that is nonzero only for read-only calls, and zero for normal slot execution, even if the execution costs gas
* in the future it will allow us to know how much gas is left at the end of a block execution, which can be filled-up with extra autonomous SCs

Reply:
another option is to not add gas_cost to ExecutionOutput at all, but to allow execute_readonly_request to return a pair (execution_output, used_gas)